File: python/baekjoon/BOJ17471게리맨더링.py
from sys import stdin
N = int(stdin.readline())
population = list(map(int, stdin.readline().split()))
arr = []
total = 0
result = 1e10
check = [0] * N

for i in range(N):
    info = list(map(int, stdin.readline().split()))
    info = list(map(lambda x: x-1, info))
    arr.append(info[1:])

total = sum(population)


def dfs(x, mask, area):
    global check
    for i in arr[x]:
        if area == 1 and check[i] == 0 and mask & 1 << i != 0:
            check[i] = area
            dfs(i, mask, area)
        elif area == 2 and check[i] == 0 and (mask & 1 << i) == 0:
            check[i] = area
            dfs(i, mask, area)


def comb(depth, mask, start, area_1):
    global check
    global result

    if depth >= 1 and depth <= N//2:
        check = [0] * N

        for i in range(N):
            if mask & 1 << i:
                check[i] = 1
                dfs(i, mask, 1)
                break

        for i in range(N):
            if mask & 1 << i == 0:
                check[i] = 2
                dfs(i, mask, 2)
                break

        if 0 not in check:
            result = min(result, abs(total-area_1*2))

    for i in range(start, N):
        if (mask & 1 << i) != 0:
            continue
        comb(depth+1, mask | 1 << i, i, area_1+population[i])


comb(0, 0, 0, 0)
print(-1 if result == 1e10 else result)

File: java/baekjoon/BOJ17471게리맨더링.java
package add;

import java.util.*;
import java.io.*;

public class BOJ17471게리맨더링 {

	static int N, total, min = Integer.MAX_VALUE, cnt;
	static int[] people, check;
	static ArrayList<Integer>[] edge;
	
	public static void combination(int depth, int mask, int start) {
		if(depth > N/2) return;
		if(depth >= 1 && depth <= N/2) {
			total = 0;
			cnt = 0;
			check = new int[N];
			boolean one = false;
			boolean two = false;
			for(int i=0; i<N; i++) {
				if(((1<<i) & mask) != 0 && !one) {
					one = true;
					total += people[i];
					dfs(i,1, mask);
				}
				else if (((1<<i) & mask) == 0 && !two){
					two = true;
					total -= people[i];
					dfs(i,2, mask);
				}
			}
			
			if(cnt != N) total = Integer.MAX_VALUE;
			min = Math.min(min, Math.abs(total));
			
		}
		
		for(int i=start; i<N; i++) {
			if(((1<<i) & mask) == 0) {
				combination(depth+1, mask | 1<<i, i);
			}
		}
	}
	
	private static void dfs(int x, int area, int mask) {
		check[x] = area;
		cnt++;
		for(int v : edge[x]) {
			if(area == 1 && check[v] == 0 && ((mask & 1<<v) != 0)) {
				total += people[v];
				dfs(v, area, mask);
			}
			else if(area == 2 && check[v] == 0 && ((mask & 1<<v) == 0)) {
				total -= people[v];
				dfs(v, area, mask);
			}
		}
	}

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		people = new int[N];
		StringTokenizer st = new StringTokenizer(br.readLine());
		edge = new ArrayList[N];
		for (int i = 0; i < N; i++) {
			people[i] = Integer.parseInt(st.nextToken());
			edge[i] = new ArrayList();
		}
		
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			int n = Integer.parseInt(st.nextToken());
			for (int j = 0; j < n; j++) {
				int x = Integer.parseInt(st.nextToken()) - 1;
				edge[i].add(x);
				edge[x].add(i);
			}
		}

		combination(0,0,0);
		System.out.println(min==Integer.MAX_VALUE ? -1 : min);
		
	}
}
